enterprise-suite/es-grafana/kubernetes-exporter.json

```diff
@@ -1,7 +1,38 @@
 [
-   {"graphName": "CPU Percent", "promQL":["node_cpu_percent"]},
-   {"graphName": "Memory Available Percent", "promQL":["node_memory_usable_percent"]},
-   {"graphName": "File System Free Percent", "promQL":["node_filesystem_free_percent"]},
-   {"graphName": "Bytes Received Per Second", "promQL":["sum by (instance) (irate(node_network_receive_bytes[1m]))"]},
-   {"graphName": "Bytes Transmitted Per Second", "promQL":["sum by (instance) (irate(node_network_transmit_bytes[1m]))"]}
+    {
+        "graphName": "CPU Percent",
+        "promQL": [
+            "sum without(container_name,id,image,name)((irate(container_cpu_user_seconds_total{ContextTags}[1m]) + irate(container_cpu_system_seconds_total[1m])) * 100)"
+        ]
+    },
+    {
+        "graphName": "Active Memory Used Percent",
+        "promQL": [
+            "sum without(container_name,id,image,name)((container_memory_working_set_bytes{ContextTags} / container_spec_memory_limit_bytes < +Inf or container_memory_working_set_bytes{ContextTags} / on (node_ip) group_left node_memory_MemTotal) * 100)"
```

Comment:
I think RSS is far more useful than working_set_bytes. People want to know how close their pod is to being OOMKilled - RSS will tell you that. working_set_bytes - I'm not sure to be honest, it will be useful in some narrow contexts for perf analysis. With a JVM process I imagine working_set_bytes will be close to RSS always anyways.

Can we have RSS here instead? Or include both of them? RSS + file caches are the most important memory metrics on a Linux system for monitoring. It also answers the question "Why is my total usage larger than my RSS?" - because RSS + file_cache = total usage.

Reply:
I'm happy to use RSS here, but just want to pause on the rationale.

Rationale: I added this based on your good suggestion to monitor for container pressure (when will I be oom'd). Now I haven't worked in kernel vm code for a long time, but from what little I do know, predicting when the kernel will be unable to fulfill a slab request is crazy hard. It is also my understanding that "working_set" is an attempt to do better than RSS for this very purpose, as it better captures the "pressure" aspect, meaning it measures what has the kernel not been able to put down, what is actively on wire, regardless of purpose. In this circumstance, slab allocations over the remainder will be at least non-trivial to handle and very likely to oom.

In any event, setting aside the theory, the kubelet itself is using working_set as the high water line for kube memory pressure / eviction triggers:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/memory_threshold_notifier.go#L80-L112

That's the same data point we're collecting, and the reason for using it.

Anwyay, the idea of using this is to more accurately reflect the OOM line and not potentially underestimate it by only considering RSS.

Reply:
more evidence, FYI:

> The better metric is container_memory_working_set_bytes as this is what the OOM killer is watching for.

https://blog.freshtracks.io/a-deep-dive-into-kubernetes-metrics-part-3-container-resource-metrics-361c5ee46e66

> The kubelet exludes inactive_file (i.e. # of bytes of file-backed memory on inactive LRU list) from its calculation as it assumes that memory is reclaimable under pressure.

https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/

> script reproduces what the kubelet does... mem_avail = capacity - working_set

https://github.com/kubernetes/website/pull/2892/files#diff-e8b9869d9d8022398a9f32d68250d792

Reply:
and here's a bunch of folks "wishing it wasn't so"

https://github.com/kubernetes/kubernetes/issues/43916

although I think they're perhaps up against other problems and getting confused.

Anyway, this is the line we should be watching, pretty sure!

Reply:
https://blog.freshtracks.io/a-deep-dive-into-kubernetes-metrics-part-3-container-resource-metrics-361c5ee46e66 is just wrong I think. The OOMKiller considers reclaimable memory. On a system without swap, the only reclaimable pages are file system cache - so RSS is the value to look at. I'm not even sure OOMKiller will consider inactive anon pages - it takes too long to swap to disk.

Working set bytes _might_ (big IF) be used if disk swap is available - the kernel can then swap out those unused pages under memory pressure. But no reasonable production system should have swap on. Especially with Kubernetes! In the vast majority of cases, working set bytes is going to be almost meaningless, unless it happens to be close to RSS.

Reply:
I fully understand the difference between reclaimable pages vs anon pages btw. So I agree with all the links you provide, except for that first blogpost. We shouldn't count filesystem pages towards memory pressure (with some exceptions, like databases which heavily rely on fs caches).

My issue is with this cadvisor metric working_memory_bytes - which has _nothing_ to do with the OOMKiller or actual physical memory (anon pages) used as far as I can tell. Instead, it looks like some kernel book keeping of recently used pages in the process. I've never seen anyone use this in actual monitoring, because as far as I can tell, it's not easily available in /proc either. I'm not actually sure where cadvisor is getting this number from. I can probably dig a bit more into the kernel code / cadvisor to see what exactly this value is used for. It might be useful in some limited contexts, but RSS and file caches are the important values.

Reply:
Okay, I did the digging.

You can see how cadvisor calculates this value (which is not from Linux, so has nothing to do with OOMKiller - it's cAdvisor's own creation): https://github.com/google/cadvisor/blob/master/container/libcontainer/handler.go#L530

It's calculated as:

     total_usage - total_inactive_file

`total_inactive_file` is the sum of `inactive_file` for the process. `inactive_file` is defined at https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt:

    inactive_file	- # of bytes of file-backed memory on inactive LRU list.

So this cadvisor metric working_set_bytes means: "Total memory usage, not including inactive file system cache". So it's actually:

    working_set_bytes = rss + total_active_file

So it's showing resident memory and all the recently used file system pages. That is actually a pretty nice metric - it's showing the "real" memory profile. For something like prometheus, this value will probably include almost all of the filesystem cache. For something that is JVM based (and not a db), it will probably exclude almost all of the filesystem cache.

So it's a useful metric. But it's a bit fuzzier than the other metrics, like RSS and file_cache. My conclusion is:

* RSS will show you the actual physical page usage that is not reclaimable. If RSS exceeds your memory limit, you get OOMKilled.
* file_cache shows all the reclaimable pages. The OS will reclaim all of these before the OOMKiller is triggered.
* cAdvisor's working_memory_bytes is a derived value. It gives an idea of the amount of "extra" memory needed beyond anon pages, to keep your process performing well. This would be very useful with any database that relies heavily on the file_cache.

I think RSS+file_cache are still the most important to monitor, working_memory_bytes will be useful as a performance metric (keep your total memory limit > working_set_bytes for best performance).

Reply:
So  I completely missed in David's comment that kubelet is actually using working_set_bytes as part of its eviction calculations. So if a user enables kubelet eviction policies (https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#eviction-policy), they'll run into this.